pkg/controllers/override/imageparse.go

```diff
@@ -0,0 +1,175 @@
+func (i *Image) OperateRegistry(operator, value string) {
+	switch operator {
+	case OperatorAdd:
+		i.registry += value
+	case OperatorReplace:
+		i.registry = value
+	case OperatorRemove:
+		i.registry = ""
+	}
+}
+
+func (i *Image) OperateRepository(operator, value string) {
+	switch operator {
+	case OperatorAdd:
+		i.repository += value
+	case OperatorReplace:
+		i.repository = value
+	case OperatorRemove:
+		i.repository = ""
+	}
+}
+
+func (i *Image) OperateTag(operator, value string) {
+	var newTag string
+	switch operator {
+	case OperatorAdd:
+		newTag = i.tag + value
+	case OperatorReplace:
+		newTag = value
+	case OperatorRemove:
+		newTag = ""
+	}
+
+	if operator != OperatorRemove && !utilreference.AnchoredTagRegexp.MatchString(newTag) {
+		return
+	}
+
+	i.tag = newTag
+}
+
+func (i *Image) OperateDigest(operator, value string) {
+	var newDigest string
+	switch operator {
+	case OperatorAdd:
+		newDigest = i.digest + value
+	case OperatorReplace:
+		newDigest = value
+	case OperatorRemove:
+		newDigest = ""
+	}
+
+	if operator != OperatorRemove && !utilreference.AnchoredDigestRegexp.MatchString(newDigest) {
+		return
+	}
+
+	i.digest = newDigest
+}
+
+func (i *Image) String() string {
+	fullName := i.repository
+
+	if i.registry != "" {
+		fullName = strings.Join([]string{i.registry, i.repository}, "/")
+	}
+
+	if i.tag != "" {
+		fullName = strings.Join([]string{fullName, i.tag}, ":")
+	}
+
+	if i.digest != "" {
+		fullName = strings.Join([]string{fullName, i.digest}, "@")
+	}
+
+	return fullName
+}
+
+// ParseImage returns a Components of the given image.
+func ParseImage(fullName string) (*Image, error) {
+	ref, err := reference.Parse(fullName)
+	if err != nil {
+		return nil, err
+	}
+
+	image := &Image{}
+	if named, ok := ref.(reference.Named); ok {
+		image.registry, image.repository = SplitRegistryAndRepository(named.Name())
+	}
+
+	if tagged, ok := ref.(reference.Tagged); ok {
+		image.tag = tagged.Tag()
+	}
+
+	if digested, ok := ref.(reference.Digested); ok {
+		image.digest = digested.Digest().String()
+	}
+
+	return image, nil
+}
+
+func SplitRegistryAndRepository(name string) (registry, repository string) {
+	i := strings.IndexRune(name, '/')
+	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != "localhost") {
+		registry, repository = "", name
+	} else {
+		registry, repository = name[:i], name[i+1:]
+	}
+	return
+}
+
+func ParseValueFromUnstructuredObj(input *unstructured.Unstructured, path string) (string, error) {
+	segments := strings.Split(strings.Trim(path, pathSeparator), pathSeparator)
+	currentObj := input.Object
+
+	for i := 0; i < len(segments)-1; i++ {
+		segment := segments[i]
+		switch obj := currentObj[segment].(type) {
+		case map[string]interface{}:
+			currentObj = obj
+		case []interface{}:
+			index, err := strconv.ParseInt(segments[i+1], 10, 32)
+			if err != nil {
+				return "", fmt.Errorf("segment(%s) of path is not a number", segments[i+1])
+			}
+			if int(index) >= len(obj) {
```

Comment:
Can index be negative?

Reply:
The code has been updated here to check if the index is a negative number.